File: src_bak/tizenclaw/channel/mcp_server.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "mcp_server.hh"

#include <dirent.h>
#include <unistd.h>

#include <fstream>
#include <iostream>

#include "../../common/logging.hh"
#include "../core/agent_core.hh"
#include "../core/skill_manifest.hh"

#include <filesystem>

namespace tizenclaw {

McpServer::McpServer(AgentCore* agent) : agent_(agent) { DiscoverTools(); }

void McpServer::DiscoverTools() {
  tools_.clear();

  // Scan skill manifests (SKILL.md > manifest.json)
  const std::string skills_dir = "/opt/usr/share/tizen-tools/skills";

  namespace fs = std::filesystem;
  std::error_code ec;
  if (fs::is_directory(skills_dir, ec)) {
    for (const auto& entry :
         fs::directory_iterator(skills_dir, ec)) {
      if (!entry.is_directory()) continue;
      auto name = entry.path().filename().string();
      if (name[0] == '.' || name == "mcp_server")
        continue;

      nlohmann::json j =
          SkillManifest::Load(entry.path().string());
      if (j.empty()) continue;

      try {
        if (j.contains("parameters")) {
          ToolInfo t;
          t.name = j.value("name", name);
          t.description = j.value("description", "");
          t.input_schema = j["parameters"];
          t.is_skill = true;
          tools_.push_back(t);
          LOG(INFO) << "MCP: Discovered tool: " << t.name;
        }
      } catch (...) {
        LOG(WARNING) << "MCP: Failed to parse: "
                     << entry.path().string();
      }
    }
  }

  // Add synthetic tool: ask_tizenclaw
  ToolInfo ask_tool;
  ask_tool.name = "ask_tizenclaw";
  ask_tool.description =
      "Send a natural language prompt to the "
      "TizenClaw AI Agent. The agent will plan "
      "and execute actions using available tools "
      "to fulfill the request.";
  ask_tool.input_schema = {{"type", "object"},
                           {"properties",
                            {{"prompt",
                              {{"type", "string"},
                               {"description",
                                "The user's request in natural "
                                "language"}}}}},
                           {"required", nlohmann::json::array({"prompt"})}};
  ask_tool.is_skill = false;
  tools_.push_back(ask_tool);

  LOG(INFO) << "MCP: Total tools discovered: " << tools_.size();
}

void McpServer::RunStdio() {
  LOG(INFO) << "MCP Server started (stdio mode)";

  // Read line-delimited JSON-RPC from stdin
  std::string line;
  while (std::getline(std::cin, line)) {
    if (line.empty()) continue;

    try {
      auto request = nlohmann::json::parse(line);
      auto response = ProcessRequest(request);

      if (!response.is_null()) {
        std::cout << response.dump() << "\n";
        std::cout.flush();
      }
    } catch (const std::exception& e) {
      LOG(ERROR) << "MCP: JSON parse error: " << e.what();
      // Send JSON-RPC parse error
      nlohmann::json err_resp = {
          {"jsonrpc", "2.0"},
          {"id", nullptr},
          {"error", {{"code", -32700}, {"message", "Parse error"}}}};
      std::cout << err_resp.dump() << "\n";
      std::cout.flush();
    }
  }

  LOG(INFO) << "MCP Server stdio loop ended";
}

nlohmann::json McpServer::ProcessRequest(const nlohmann::json& request) {
  std::string method = request.value("method", "");
  auto params = request.value("params", nlohmann::json::object());
  auto req_id = request.value("id", nlohmann::json(nullptr));

  nlohmann::json result;

  if (method == "initialize") {
    result = HandleInitialize(params);
  } else if (method == "notifications/initialized") {
    // Notification — no response needed
    return nlohmann::json();
  } else if (method == "tools/list") {
    result = HandleToolsList(params);
  } else if (method == "tools/call") {
    result = HandleToolsCall(params, STDOUT_FILENO);
  } else {
    // Unknown method
    nlohmann::json response = {
        {"jsonrpc", "2.0"},
        {"id", req_id},
        {"error", {{"code", -32601}, {"message", "Method not found"}}}};
    return response;
  }

  nlohmann::json response = {
      {"jsonrpc", "2.0"}, {"id", req_id}, {"result", result}};
  return response;
}

nlohmann::json McpServer::HandleInitialize(const nlohmann::json& /*params*/) {
  return {{"protocolVersion", kProtocolVersion},
          {"capabilities", {{"tools", nlohmann::json::object()}}},
          {"serverInfo",
           {{"name", "TizenClaw-MCP-Server"}, {"version", kVersion}}}};
}

nlohmann::json McpServer::HandleToolsList(const nlohmann::json& /*params*/) {
  nlohmann::json tools_array = nlohmann::json::array();

  for (auto& t : tools_) {
    tools_array.push_back({{"name", t.name},
                           {"description", t.description},
                           {"inputSchema", t.input_schema}});
  }

  return {{"tools", tools_array}};
}

nlohmann::json McpServer::HandleToolsCall(const nlohmann::json& params,
                                          int /*stdout_fd*/) {
  std::string tool_name = params.value("name", "");
  auto arguments = params.value("arguments", nlohmann::json::object());

  // Find the tool
  const ToolInfo* found = nullptr;
  for (auto& t : tools_) {
    if (t.name == tool_name) {
      found = &t;
      break;
    }
  }

  if (!found) {
    return {
        {"isError", true},
        {"content",
         nlohmann::json::array(
             {{{"type", "text"}, {"text", "Tool not found: " + tool_name}}})}};
  }

  LOG(INFO) << "MCP: Calling tool: " << tool_name;

  if (!found->is_skill) {
    // ask_tizenclaw: route through Agentic Loop
    std::string prompt = arguments.value("prompt", "");
    if (prompt.empty()) {
      return {
          {"isError", true},
          {"content",
           nlohmann::json::array(
               {{{"type", "text"}, {"text", "Missing 'prompt' argument"}}})}};
    }

    std::string result = agent_->ProcessPrompt("mcp_session", prompt);

    return {{"content",
             nlohmann::json::array({{{"type", "text"}, {"text", result}}})}};
  }

  // Direct skill execution via container
  std::string result = agent_->ExecuteSkillForMcp(tool_name, arguments);

  return {{"content",
           nlohmann::json::array({{{"type", "text"}, {"text", result}}})}};
}

}  // namespace tizenclaw

File: src/tizenclaw/src/channel/mcp_server.rs
//! MCP Server — Anthropic Model Context Protocol (JSON-RPC 2.0 over stdio).
//!
//! Implements the MCP server protocol for exposing TizenClaw tools
//! to external MCP clients (e.g., Claude Desktop, Cursor).
//!
//! Protocol version: 2024-11-05
//! Transport: stdio (line-delimited JSON-RPC)

use serde_json::{json, Value};
use crate::core::skill_manifest::SkillManifest;

const PROTOCOL_VERSION: &str = "2024-11-05";
const SERVER_VERSION: &str = "1.0.0";
const SKILLS_DIR: &str = "/opt/usr/share/tizen-tools/skills";

/// A tool discovered from SKILL.md or built-in.
struct McpToolInfo {
    name: String,
    description: String,
    input_schema: Value,
    is_skill: bool,
}

pub struct McpServer {
    tools: Vec<McpToolInfo>,
}

impl McpServer {
    pub fn new() -> Self {
        let mut server = McpServer { tools: Vec::new() };
        server.discover_tools();
        server
    }

    /// Scan `/opt/usr/share/tizen-tools/skills` for SKILL.md manifests
    /// and register them as MCP tools.
    fn discover_tools(&mut self) {
        self.tools.clear();

        // Scan skill manifests
        if let Ok(entries) = std::fs::read_dir(SKILLS_DIR) {
            for entry in entries.flatten() {
                let path = entry.path();
                if !path.is_dir() {
                    continue;
                }

                let dir_name = path.file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or("");
                if dir_name.starts_with('.') || dir_name == "mcp_server" {
                    continue;
                }

                if let Some(manifest) = SkillManifest::load(&path) {
                    log::info!("MCP: Discovered tool: {}", manifest.name);
                    self.tools.push(McpToolInfo {
                        name: manifest.name,
                        description: manifest.description,
                        input_schema: manifest.input_schema,
                        is_skill: true,
                    });
                }
            }
        }

        // Add synthetic tool: ask_tizenclaw
        self.tools.push(McpToolInfo {
            name: "ask_tizenclaw".into(),
            description: "Send a natural language prompt to the TizenClaw AI Agent. \
                         The agent will plan and execute actions using available tools \
                         to fulfill the request.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "prompt": {
                        "type": "string",
                        "description": "The user's request in natural language"
                    }
                },
                "required": ["prompt"]
            }),
            is_skill: false,
        });

        log::info!("MCP: Total tools discovered: {}", self.tools.len());
    }

    /// Run the stdio JSON-RPC 2.0 loop (blocking).
    ///
    /// Reads line-delimited JSON from stdin, writes responses to stdout.
    /// This is intended to be called when the daemon is launched with
    /// `--mcp-stdio` flag.
    pub fn run_stdio<F>(&self, process_prompt: F)
    where
        F: Fn(&str, &str) -> String,  // (session_id, prompt) -> result
    {
        log::info!("MCP Server started (stdio mode)");
        use std::io::BufRead;

        let stdin = std::io::stdin();
        for line in stdin.lock().lines() {
            let line = match line {
                Ok(l) => l,
                Err(_) => break,
            };
            if line.is_empty() {
                continue;
            }

            let request: Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(e) => {
                    log::error!("MCP: JSON parse error: {}", e);
                    let err_resp = json!({
                        "jsonrpc": "2.0",
                        "id": null,
                        "error": {"code": -32700, "message": "Parse error"}
                    });
                    println!("{}", err_resp);
                    continue;
                }
            };

            let response = self.process_request(&request, &process_prompt);
            if !response.is_null() {
                println!("{}", response);
            }
        }

        log::info!("MCP Server stdio loop ended");
    }

    /// Process a single JSON-RPC 2.0 request.
    fn process_request<F>(&self, request: &Value, process_prompt: &F) -> Value
    where
        F: Fn(&str, &str) -> String,
    {
        let method = request["method"].as_str().unwrap_or("");
        let params = request.get("params").cloned().unwrap_or_else(|| json!({}));
        let req_id = request.get("id").cloned().unwrap_or(Value::Null);

        match method {
            "initialize" => {
                let result = self.handle_initialize();
                json!({"jsonrpc": "2.0", "id": req_id, "result": result})
            }
            "notifications/initialized" => {
                // Notification — no response
                Value::Null
            }
            "tools/list" => {
                let result = self.handle_tools_list();
                json!({"jsonrpc": "2.0", "id": req_id, "result": result})
            }
            "tools/call" => {
                let result = self.handle_tools_call(&params, process_prompt);
                json!({"jsonrpc": "2.0", "id": req_id, "result": result})
            }
            _ => {
                json!({
                    "jsonrpc": "2.0",
                    "id": req_id,
                    "error": {"code": -32601, "message": "Method not found"}
                })
            }
        }
    }

    fn handle_initialize(&self) -> Value {
        json!({
            "protocolVersion": PROTOCOL_VERSION,
            "capabilities": {"tools": {}},
            "serverInfo": {
                "name": "TizenClaw-MCP-Server",
                "version": SERVER_VERSION
            }
        })
    }

    fn handle_tools_list(&self) -> Value {
        let tools: Vec<Value> = self.tools.iter().map(|t| {
            json!({
                "name": t.name,
                "description": t.description,
                "inputSchema": t.input_schema
            })
        }).collect();

        json!({"tools": tools})
    }

    fn handle_tools_call<F>(&self, params: &Value, process_prompt: &F) -> Value
    where
        F: Fn(&str, &str) -> String,
    {
        let tool_name = params["name"].as_str().unwrap_or("");
        let arguments = params.get("arguments").cloned().unwrap_or_else(|| json!({}));

        let found = self.tools.iter().find(|t| t.name == tool_name);
        let tool = match found {
            Some(t) => t,
            None => {
                return json!({
                    "isError": true,
                    "content": [{"type": "text", "text": format!("Tool not found: {}", tool_name)}]
                });
            }
        };

        log::info!("MCP: Calling tool: {}", tool_name);

        if !tool.is_skill {
            // ask_tizenclaw: route through agentic loop
            let prompt = arguments["prompt"].as_str().unwrap_or("");
            if prompt.is_empty() {
                return json!({
                    "isError": true,
                    "content": [{"type": "text", "text": "Missing 'prompt' argument"}]
                });
            }

            let result = process_prompt("mcp_session", prompt);
            return json!({
                "content": [{"type": "text", "text": result}]
            });
        }

        // Direct skill execution
        let result = process_prompt("mcp_skill", &format!("[SKILL:{}] {}", tool_name, arguments));
        json!({
            "content": [{"type": "text", "text": result}]
        })
    }
}
